storage/filestore/store.go

```diff
@@ -90,3 +158,33 @@ func (store *Store) FreeSpace() (int64, error) {
 	}
 	return info.AvailableSpace, nil
 }
+
+// GetAllNamespaces finds all known namespace IDs in use in local storage. They are not
+// guaranteed to contain any blobs.
+func (store *Store) GetAllNamespaces(ctx context.Context) (ids [][]byte, err error) {
+	return store.dir.GetAllNamespaces(ctx)
+}
+
+// ForAllKeysInNamespace executes doForEach for each locally stored blob in the given
+// namespace. If doForEach returns a non-nil error, ForAllKeysInNamespace will stop
+// iterating and return the error immediately.
+func (store *Store) ForAllKeysInNamespace(ctx context.Context, namespace []byte, doForEach func(storage.StoredBlobAccess) error) (err error) {
+	return store.dir.ForAllKeysInNamespace(ctx, namespace, doForEach)
+}
+
+// StoreForTest is a wrapper for Store that also allows writing new V0 blobs (in order to test
+// situations involving those)
+type StoreForTest struct {
```

Comment:
Why does the test store live in the `store.go` file instead of being in the `store_test.go` file?

Reply:
This isn't a test, it's just a wrapper around `Store` that can be used by tests (so that we don't need to have `CreateV0` etc. on the normal `Store` type).

store_test.go can't just define the wrapper type itself because it uses a separate package- `filestore_test`, which helps to ensure the right things are exported.

Reply:
But isn't it only being used in test files, and only exists for testing purposes? In general having test-specific structs exist in non-test files feels like a code smell, a chance to redesign some things to make them more testable?

Reply:
I strongly disagree that test-specific code in non-test files is a code smell. It is non-desirable, of _course_, but often the alternative is using mocks. Mocks _are_ a code smell. When it comes to a choice between test-specific hooks into non-test code, and incomplete or broken tests, I will always choose the former.

I doubt you could redesign the tests that use `CreateV0` so that it became unnecessary to have any test-only code in here, without either crippling the tests or exporting things that shouldn't be exported.

Reply:
You are much more knowledgeable about this part of the code, I'm happy to defer to your judgment on this.

Reply:
For these situations I often end up using a `fs := map[string]string{` for the initial state of the filesystem and then run a test based on that folder. This also ensures that backcomp doesn't change, even if you accidentally change `CreateV0`.

I think here, the test code isn't too problematic, since it doesn't import any testing specific code.

Reply:
That is a good idea Egon, and you're right, it would remove the necessity for this test-specific code. I think we may have to put it off until after this giant kidney stone of a PR can be gotten through.